moveit_core/robot_state/src/cartesian_interpolator.cpp

```diff
@@ -189,147 +295,73 @@ CartesianInterpolator::Percentage CartesianInterpolator::computeCartesianPath(
-                                                                             std::vector<RobotStatePtr>& traj,
+                                                                             std::vector<RobotStatePtr>& path,
                                                                              const JumpThreshold& jump_threshold)
 {
-  double percentage_solved = 1.0;
-  if (traj.size() <= 1)
-    return percentage_solved;
-
-  if (jump_threshold.factor > 0.0)
-    percentage_solved *= checkRelativeJointSpaceJump(group, traj, jump_threshold.factor);
+  std::optional<int> jump_index = hasJointSpaceJumps(path, group, jump_threshold);
 
-  if (jump_threshold.revolute > 0.0 || jump_threshold.prismatic > 0.0)
-    percentage_solved *= checkAbsoluteJointSpaceJump(group, traj, jump_threshold.revolute, jump_threshold.prismatic);
+  double percentage_solved = 1.0;
+  if (jump_index)
+  {
+    percentage_solved = static_cast<double>(*jump_index) / static_cast<double>(path.size());
+    // Truncate the path at the index right before the jump.
+    path.resize(*jump_index);
+  }
 
   return CartesianInterpolator::Percentage(percentage_solved);
 }
 
-CartesianInterpolator::Percentage CartesianInterpolator::checkRelativeJointSpaceJump(const JointModelGroup* group,
-                                                                                     std::vector<RobotStatePtr>& traj,
-                                                                                     double jump_threshold_factor)
+std::optional<int> hasJointSpaceJumps(const std::vector<moveit::core::RobotStatePtr>& waypoints,
+                                      const moveit::core::JointModelGroup* group,
+                                      const moveit::core::JumpThreshold& jump_threshold)
 {
-  if (traj.size() < MIN_STEPS_FOR_JUMP_THRESH)
+  if (waypoints.size() <= 1)
   {
-    RCLCPP_WARN(LOGGER,
-                "The computed trajectory is too short to detect jumps in joint-space "
-                "Need at least %zu steps, only got %zu. Try a lower max_step.",
-                MIN_STEPS_FOR_JUMP_THRESH, traj.size());
+    return std::nullopt;
   }
 
-  std::vector<double> dist_vector;
-  dist_vector.reserve(traj.size() - 1);
-  double total_dist = 0.0;
-  for (std::size_t i = 1; i < traj.size(); ++i)
+  if (jump_threshold.relative_factor > 0.0)
   {
-    double dist_prev_point = traj[i]->distance(*traj[i - 1], group);
-    dist_vector.push_back(dist_prev_point);
-    total_dist += dist_prev_point;
+    return hasRelativeJointSpaceJump(waypoints, group, jump_threshold.relative_factor);
   }
 
-  double percentage = 1.0;
-  // compute the average distance between the states we looked at
-  double thres = jump_threshold_factor * (total_dist / static_cast<double>(dist_vector.size()));
-  for (std::size_t i = 0; i < dist_vector.size(); ++i)
+  if (jump_threshold.revolute > 0.0 || jump_threshold.prismatic > 0.0)
```

Comment:
I think it is possible to misconfigure this struct that both jump types are configured. Can we prevent this somehow (consider making the members private) or print a warning if that's the case?

Reply:
Yeah, the original intent of this change was to prevent incorrect initialization of the object. I originally had the constructors private to force users to use the named builders. But unfortunately that made MTC fail to build in CI (relies on the current API). So I made the constructors public again to not break MTC, but we should consider updating MTC next, and come back to this after that.
That being said I would also not invest much more time into these APIs, since much of this will be eventually replaced with the new path IK.